sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/Storage/BlobsCheckpointStore.cs

```diff
@@ -247,37 +308,130 @@ await foreach (BlobItem blob in ContainerClient.GetBlobsAsync(traits: BlobTraits
+            var blobClient = ContainerClient.GetBlobClient(blobName);
 
             var metadata = new Dictionary<string, string>()
-                {
-                    { BlobMetadataKey.Offset, checkpoint.Offset.ToString() },
-                    { BlobMetadataKey.SequenceNumber, checkpoint.SequenceNumber.ToString() }
-                };
+            {
+                { BlobMetadataKey.Offset, checkpoint.Offset.ToString() },
+                { BlobMetadataKey.SequenceNumber, checkpoint.SequenceNumber.ToString() }
+            };
+
+            Func<CancellationToken, Task> updateCheckpointAsync = async updateCheckpointToken =>
+            {
+                using var blobContent = new MemoryStream(new byte[0]);
+                await blobClient.UploadAsync(blobContent, metadata: metadata, cancellationToken: updateCheckpointToken);
+            };
 
-            MemoryStream blobContent = null;
             try
             {
-                blobContent = new MemoryStream(new byte[0]);
-                await blobClient.UploadAsync(blobContent, metadata: metadata).ConfigureAwait(false);
+                return ApplyRetryPolicy(updateCheckpointAsync, cancellationToken);
                 // TODO: Add log  - "Checkpoint with partition id = '{ checkpoint.PartitionId }' updated."
             }
             catch (RequestFailedException ex) when (ex.ErrorCode == BlobErrorCode.ContainerNotFound)
             {
                 // TODO: Add log  - "Checkpoint with partition id = '{ checkpoint.PartitionId }' could not be updated because specified container does not exist."
                 throw new RequestFailedException(Resources.BlobsResourceDoesNotExist);
             }
-            finally
+        }
+
+        /// <summary>
+        ///   Applies the checkpoint store's <see cref="RetryPolicy" /> to a specified function.
+        /// </summary>
+        ///
+        /// <param name="functionToRetry">The function to which the retry policy should be applied.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>The value returned by the function to which the retry policy has been applied.</returns>
+        ///
+        private async Task ApplyRetryPolicy(Func<CancellationToken, Task> functionToRetry,
+                                            CancellationToken cancellationToken)
+        {
+            var failedAttemptCount = 0;
+            var retryDelay = default(TimeSpan?);
+            var tryTimeout = RetryPolicy.CalculateTryTimeout(0);
+
+            var stopWatch = Stopwatch.StartNew();
+
+            while (!cancellationToken.IsCancellationRequested)
             {
-                blobContent?.Dispose();
+                try
+                {
+                    var timeoutToken = (new CancellationTokenSource(tryTimeout)).Token;
```

Comment:
`CancellationTokenSource` is disposable.  You're not able to dispose properly this way.   I'd recommend hoisting the source into its own `using` declaration rather than making it a compound assignment.